lib/core/cmap/message_stream.js

```diff
@@ -0,0 +1,177 @@
+const BufferList = require('bl');
+const MongoParseError = require('../error').MongoParseError;
+const decompress = require('../wireprotocol/compression').decompress;
+const Response = require('../connection/commands').Response;
+const BinMsg = require('../connection/msg').BinMsg;
+const MongoError = require('../error').MongoError;
+const OP_COMPRESSED = require('../wireprotocol/shared').opcodes.OP_COMPRESSED;
+const OP_MSG = require('../wireprotocol/shared').opcodes.OP_MSG;
+const MESSAGE_HEADER_SIZE = require('../wireprotocol/shared').MESSAGE_HEADER_SIZE;
+const COMPRESSION_DETAILS_SIZE = require('../wireprotocol/shared').COMPRESSION_DETAILS_SIZE;
+const opcodes = require('../wireprotocol/shared').opcodes;
+const compress = require('../wireprotocol/compression').compress;
+const compressorIDs = require('../wireprotocol/compression').compressorIDs;
+const uncompressibleCommands = require('../wireprotocol/compression').uncompressibleCommands;
+const Msg = require('../connection/msg').Msg;
+
+const kDefaultMaxBsonMessageSize = 1024 * 1024 * 16 * 4;
+const kBuffer = Symbol('buffer');
+
+/**
+ * A duplex stream that is capable of reading and writing raw wire protocol messages, with
+ * support for optional compression
+ */
+class MessageStream extends Duplex {
+  constructor(options) {
+    options = options || {};
+    super(options);
+
+    this.bson = options.bson;
+    this.maxBsonMessageSize = options.maxBsonMessageSize || kDefaultMaxBsonMessageSize;
+
+    this[kBuffer] = new BufferList();
+  }
+
+  _write(chunk, _, callback) {
+    this[kBuffer].append(chunk);
+
+    while (this[kBuffer].length >= 4) {
+      const sizeOfMessage = this[kBuffer].readInt32LE(0);
+      if (sizeOfMessage < 0) {
+        callback(new MongoParseError(`Invalid message size: ${sizeOfMessage}`));
+        return;
+      }
+
+      if (sizeOfMessage > this.maxBsonMessageSize) {
+        callback(
+          new MongoParseError(
+            `Invalid message size: ${sizeOfMessage}, max allowed: ${this.maxBsonMessageSize}`
+          )
+        );
+        return;
+      }
+
+      if (sizeOfMessage > this[kBuffer].length) {
+        callback();
+        return;
+      }
+
+      const messageBuffer = this[kBuffer].slice(0, sizeOfMessage);
+      processMessage(this, messageBuffer, callback);
+      this[kBuffer].consume(sizeOfMessage);
+    }
+  }
+
+  _read(/* size */) {
+    // NOTE: This implementation is empty because we explicitly push data to be read
+    //       when `writeMessage` is called.
+    return;
+  }
+
+  writeCommand(command, operationDescription) {
+    // TODO: agreed compressor should live in `StreamDescription`
+    const shouldCompress = operationDescription && !!operationDescription.agreedCompressor;
+    if (!shouldCompress || !canCompress(command)) {
+      this.push(Buffer.concat(command.toBin()));
+      return;
+    }
+
+    // otherwise, compress the message
+    const concatenatedOriginalCommandBuffer = Buffer.concat(command.toBind());
```

Comment:
`toBind` looks like a typo of `toBin`

Reply:
oops, good catch. Compression isn't tested yet since the connection isn't fully hooked up.